backends/dpdk/dpdkContext.cpp

```diff
@@ -0,0 +1,582 @@
+        if (params)
+            attr.params = params->clone();
+        else
+            attr.params = nullptr;
+
+        attr.constant_default_action = is_const_default_action;
+        attr.is_compiler_added_action = is_compiler_added_action;
+        attr.allowed_as_hit_action = can_be_hit_action;
+        attr.allowed_as_default_action = can_be_default_action;
+        attr.actionHandle = getNewActionHandle();
+        actionAttrMap.emplace(act->getName(), attr);
+    }
+}
+
+/* Print a single table object of match_table type in Context JSON*/
+void WriteContextJson::printTableCtxtJson(const IR::P4Table *tbl, std::ostream &out) {
+    bool hasActionProfileSelector = false;
+    auto mainTableAttr = ::get(tableAttrmap, tbl->name.originalName);
+    cstring tableName = mainTableAttr.controlName + "." + tbl->name.originalName;
+    add_space(out, 8); out << "{\n";
+    add_space(out, 12); out << "\"table_type\": \"" << mainTableAttr.tableType << "\",\n";
+    add_space(out, 12); out << "\"direction\": \"" << mainTableAttr.direction << "\",\n";
+    add_space(out, 12); out << "\"handle\": " << mainTableAttr.tableHandle << ",\n";
+    add_space(out, 12); out << "\"name\": \"" << tableName << "\",\n";
+    add_space(out, 12); out << "\"size\": ";
+
+    if (auto size = tbl->properties->getProperty("size")) {
+        out << size->value << ",\n";
+    } else {
+        // default table size for DPDK
+        out << "65536" << ",\n";
+    }
+
+    // Below empty arrays are currently required by the control plane software.
+    // May be removed in future.
+    add_space(out, 12); out << "\"stateful_table_refs\": []" << ",\n";
+    add_space(out, 12); out << "\"statistics_table_refs\": []" << ",\n";
+    add_space(out, 12); out << "\"meter_table_refs\": []" << ",\n";
+    add_space(out, 12); out << "\"p4_hidden\": "
+                            << std::boolalpha << mainTableAttr.isHidden << ",\n";
+
+    // Reference to compiler generated member table in case of action profile and action selector.
+    add_space(out, 12); out << "\"action_data_table_refs\": [";
+    const IR::P4Table *memberTable = nullptr;
+    if (structure->member_tables.count(tbl->name)) {
+        hasActionProfileSelector = true;
+        memberTable = structure->member_tables.at(tbl->name);
+        auto tableAttr = ::get(tableAttrmap, memberTable->name.originalName);
+        tableName = tableAttr.controlName + "." + memberTable->name.originalName;
+        out << "\n";
+        add_space(out, 16); out << "{\n";
+        add_space(out, 20);
+        out << "\"name\": \"" <<  tableName <<"\",\n";
+        add_space(out, 20); out << "\"handle\": " << tableAttr.tableHandle   <<"\n";
+        add_space(out, 16); out << "}\n";
+        add_space(out, 12);
+    }
+    out << "],\n";
+
+    // Reference to compiler generated group table in case of action selector
+    add_space(out, 12); out << "\"selection_table_refs\": [";
+    if (structure->group_tables.count(tbl->name)) {
```

Comment:
this code is almost the same as above. could it be factored into a separate function?